File: controllers/quiz_controller.py
from uuid import uuid4

from flask import Blueprint, jsonify, render_template, request, session

from services.quiz_service import GameError


def create_blueprint(service):
    bp = Blueprint('quiz', __name__)

    def owner():
        if 'owner' not in session:
            session['owner'] = str(uuid4())
        return session['owner']

    def body(*fields):
        data = request.get_json(silent=True)
        if not isinstance(data, dict):
            raise GameError('Envie um objeto JSON válido.')
        for field in fields:
            value = data.get(field)
            if not isinstance(value, str) or not value.strip() or len(value) > 100:
                raise GameError(f'Campo inválido: {field}.')
        return data

    @bp.errorhandler(GameError)
    def game_error(error):
        return jsonify(error=str(error)), error.status

    @bp.get('/')
    def index():
        owner()
        return render_template('index.html')

    @bp.get('/api/lessons')
    def lessons():
        owner()
        return jsonify(service.catalog())

    @bp.get('/api/game')
    def current():
        return jsonify(service.current(owner()))

    @bp.post('/api/game')
    def start():
        data = body('lesson_id', 'request_id')
        return jsonify(service.start(owner(), data['lesson_id'], data['request_id']))

    @bp.post('/api/game/answer')
    def answer():
        data = body('game_id', 'question_id', 'alternative_id')
        return jsonify(service.answer(owner(), data['game_id'], data['question_id'], data['alternative_id']))

    @bp.post('/api/game/next')
    def advance():
        data = body('game_id', 'question_id')
        return jsonify(service.advance(owner(), data['game_id'], data['question_id']))

    @bp.post('/api/game/abandon')
    def abandon():
        data = body('game_id')
        service.abandon(owner(), data['game_id'])
        return jsonify(ok=True)

    return bp

File: static/js/controllers/QuizController.js
export class QuizController {
  constructor(api, view) {
    this.api = api;
    this.view = view;
    this.game = null;
    this.busy = false;
    view.bind({
      menu: () => view.menu(), lessons: () => this.lessons(),
      help: () => view.help(), start: id => this.start(id),
      answer: id => this.answer(id), next: () => this.next(),
      exit: () => view.confirmExit(() => this.abandon()),
      replay: () => this.start(this.game.lesson_id),
    });
  }

  async init() {
    await this.run(async () => {
      try {
        this.game = await this.api.request('game');
        this.view.game(this.game);
      } catch (error) {
        if (error.status !== 410) throw error;
        this.view.welcome();
      }
    });
  }

  async run(action) {
    if (this.busy) return;
    this.busy = true;
    this.view.setBusy(true);
    this.view.clearError();
    try { await action(); }
    catch (error) {
      if (error.status === 410) {
        this.game = null;
        this.view.menu();
        this.view.error(error.message, () => this.lessons());
      } else if (error.status === 409) {
        this.view.error(error.message, () => this.init());
      } else {
        this.view.error(error.message, () => this.run(action));
      }
    } finally {
      this.busy = false;
      this.view.setBusy(false);
    }
  }

  lessons() {
    return this.run(async () => this.view.lessons(await this.api.request('lessons')));
  }

  start(lessonId) {
    const requestId = crypto.randomUUID();
    return this.run(async () => {
      this.game = await this.api.request('game', { lesson_id: lessonId, request_id: requestId });
      this.view.game(this.game);
    });
  }

  answer(alternativeId) {
    if (!this.game || this.game.feedback || this.game.completed) return;
    const payload = { game_id: this.game.game_id, question_id: this.game.question.id, alternative_id: alternativeId };
    return this.run(async () => {
      this.game = await this.api.request('game/answer', payload);
      this.view.game(this.game);
    });
  }

  next() {
    if (!this.game?.feedback || this.game.completed) return;
    const payload = { game_id: this.game.game_id, question_id: this.game.question.id };
    return this.run(async () => {
      this.game = await this.api.request('game/next', payload);
      this.view.game(this.game);
    });
  }

  abandon() {
    const payload = { game_id: this.game.game_id };
    return this.run(async () => {
      await this.api.request('game/abandon', payload);
      this.game = null;
      await this.api.request('lessons').then(data => this.view.lessons(data));
    });
  }
}
